pkg/kwok/controllers/pod_controller.go

```diff
@@ -420,6 +424,12 @@ func (c *PodController) watchResources(ctx context.Context, opt metav1.ListOptio
 				case watch.Added, watch.Modified:
 					pod := event.Object.(*corev1.Pod)
 					if c.need(pod) {
+						if event.Type == watch.Added && c.nodeMetricsHandler != nil {
+							c.nodeMetricsHandler.AddContainersCount(pod.Spec.NodeName, len(pod.Spec.Containers))
+							if err := c.nodeMetricsHandler.Publish(pod.Spec.NodeName); err != nil {
```

Comment:
The Metric should not be updated when the value changes, but when it is requested

Reply:
My understanding is that we need to update collectors so when the GET request is issued, it can pick the current state and serve it. When I looked into a kubelet code, it is basically doing the same thing:
https://github.com/kubernetes/kubernetes/blob/61d455ed1173cd89a98442adf4623a29c5681c58/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L1086

We just update collectors and let the prometheus http handler serve metrics. If we evaluate it on the http request, we would need to wrap the handler to execute evaluations which will increase latency of the GET request. Publish was meant to publish metrics to collectors, and is not dependent on the scraping interval at all. As soon as the state changes, it will update the metric collectors, and that is it.

Can you please provide an example where the handler evaluates metrics and serves them? I'm curious how it works :relaxed:

Reply:
This is awesome :100: ! I just updated the code, and thank you so much for this one!